### ISSUE TITLE
How to handle pyright's (justified) complaints regarding callable eqx.Module?

### ISSUE THREAD

**Issue by Artur-Galstyan** [patrick-kidger/equinox#1002]
Hi there,

when writing models, sometimes I want to pass a specific layer as an argument (this often happens with normalisation layers), e.g.:

```python
import equinox as eqx
import jax
from beartype.typing import Callable
from jaxtyping import Array


class BarNorm(eqx.Module):
    x: Array

    def __call__(self, y):
        return self.x + y


class Foo(eqx.Module):
    bar: eqx.Module

    def __init__(self, bar: Callable[..., eqx.Module]):
        self.bar = bar(1)

    def __call__(self, x):
        return self.bar(x) + x # <<< pyright complains here


f = Foo(bar=BarNorm)
print(f(jax.numpy.array(5))) # gives 11, because (5 + 1) + 5 = 1
```

Pyright rightfully complains that:

```
Object of type "Module" is not callable
  Attribute "__call__" is unknown
```

which is correct - not all `eqx.Module`s need to have a `__call__` function.

The only workaround I found was doing this:

```python
import abc

import equinox as eqx
import jax
from beartype.typing import Callable
from jaxtyping import Array


class CallableEqxModule(eqx.Module):
    @abc.abstractmethod
    def __call__(self, *_, **__):
        pass


class BarNorm(CallableEqxModule): # <<< but I had to change this from eqx.Module to CallableEqxModule
    x: Array

    def __call__(self, y):
        return self.x + y


class Foo(eqx.Module):
    bar: CallableEqxModule

    def __init__(self, bar: Callable[..., CallableEqxModule]):
        self.bar = bar(1)

    def __call__(self, x):
        return self.bar(x) + x


f = Foo(bar=BarNorm)
print(f(jax.numpy.array(5)))
```

I'm sure you definitely encountered this pyright complaint and I was wondering how you deal with it. I'd actually rather not use something like `CallableEqxModule` but then I have to litter my code with `# pyright: ignore`. WWYD?

**Comment by johannahaffner** [comment]
You can write 

```python
class Foo(eqx.Module):
    bar: BarNorm
```

**Comment by johannahaffner** [comment]
A bit more philosophically speaking, I find it helpful to be able to understand what the attribute does. In your case, creating an `AbstractNorm` that provides an interface that could be a solution, this could for example be implemented by `BarNorm`. `AbstractNorm` can then inherit from `equinox.Module`, without the need to specify this elsewhere - it anyway would not tell a reader of your code anything about the specific functionality the attribute implements.

You might also like [these musings](https://docs.kidger.site/equinox/pattern/).

**Comment by johannahaffner** [comment]
Hi Arthur, 

indeed the torch stuff leaves it up in the air what actually happens behind the curtain!

> [...] the user might pass in their own `FancyBarNorm`.
> [...] perhaps these abstract classes are the way to go.

Abstract classes can also be helpful to the users here, I believe. They make it clear what their implementation needs to do to work within the context of the functionality you are providing. An alternative could be to use a less restrictive `Callable[[Layer], Layer]`, where `Layer` might be a `TypeVar`, or an alias for a PyTree of arrays, or a Module... More flexible, but you pay for that by losing some clarity.

> Ahhh, I was looking for this! I couldn't find it in the menu on the left. I saw this some time in the past but it's like a hidden gem, hard to find :D

Agree that its a gem, hopefully it makes a comeback at some point :)

**Comment by lockwo** [comment]
Given equinox's support and integration with Abstract classes, I think that's definitely the way to go. But for the sake of completeness, structural subtyping would also work here (Abstract/final? What is this equinox.jl ;)?). Really, I would recommend abstract.

<details>
<summary> Example </summary>

```python
import equinox as eqx
import jax
from jaxtyping import Array
from beartype.typing import Callable
from typing import Protocol, Any


class NormCall(Protocol):
    def __call__(self, y):
        ...


class BarNorm(eqx.Module):
    x: Array

    def __call__(self, y):
        return self.x + y


class Foo(eqx.Module):
    bar: NormCall

    def __init__(self, bar_factory: Callable[..., NormCall]):
        self.bar = bar_factory(1)

    def __call__(self, x):
        return self.bar(x) + x


f = Foo(bar_factory=BarNorm)
print(f(jax.numpy.array(5)))
```

</details >
